### PR TITLE
allow the user to pass in a file to configure node affinities for blackduck

### LINKED ISSUE #146: Add ability to configure Node Affinity
Users may want to direct an application to specific nodes via node affinity.  Provide a means to configure node affinity in synopsysctl and the UI.

### PR DESCRIPTION
**Pull request by yashbhutwala**
Closes #146. 

Currently, keeping the configuration simple.  Example attached for Saas based on [cn-crd node affinities](https://github.com/blackducksoftware/cn-crd-controller/blob/master/pkg/blackduck/incluster_hub.go).

/assign @rrati @jdartigalongue

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/413#" title="Author self-approved">yashbhutwala</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [yashbhutwala]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/apps/blackduck/v1/containers/nodeAffinity.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+// getAffinitiesForSpecificPod iterates once through the user provided NodeAffinities, internally caches them into a map of format "podName":[]blackduckapi.NodeAffinity, and returns []blackduckapi.NodeAffinity for the given "podName"
+func (c *Creater) getAffinitiesForSpecificPod(podName string) []blackduckapi.NodeAffinity {
+	if affinitiesForSpecificPod, ok := affinityMap[podName]; ok {
+		return affinitiesForSpecificPod
+	}
+
+	affinitiesForSpecificPod := []blackduckapi.NodeAffinity{}
+	for _, affinity := range c.hubSpec.NodeAffinities {
+		affinityMap[affinity.PodName] = append(affinitiesForSpecificPod, affinity)
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/apps/blackduck/v1/containers/nodeAffinity.go]: That function is removed now.

File: pkg/apps/blackduck/latest/containers/nodeAffinity.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+// getAffinitiesForSpecificPod iterates once through the user provided NodeAffinities, internally caches them into a map of format "podName":[]blackduckapi.NodeAffinity, and returns []blackduckapi.NodeAffinity for the given "podName"
+func (c *Creater) getAffinitiesForSpecificPod(podName string) []blackduckapi.NodeAffinity {
+	if affinitiesForSpecificPod, ok := affinityMap[podName]; ok {
+		return affinitiesForSpecificPod
+	}
+
+	affinitiesForSpecificPod := []blackduckapi.NodeAffinity{}
+	for _, affinity := range c.hubSpec.NodeAffinities {
+		affinityMap[affinity.PodName] = append(affinitiesForSpecificPod, affinity)
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/apps/blackduck/latest/containers/nodeAffinity.go]: Function removed.

File: pkg/apps/blackduck/latest/containers/nodeAffinity.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+package containers
+
+import (
+	horizonapi "github.com/blackducksoftware/horizon/pkg/api"
+	blackduckapi "github.com/blackducksoftware/synopsys-operator/pkg/api/blackduck/v1"
+)
+
+var affTypeMap = map[string]horizonapi.AffinityType{
+	"AffinityHard": horizonapi.AffinityHard,
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/apps/blackduck/latest/containers/nodeAffinity.go]: Sure

File: pkg/api/blackduck/v1/types.go [hunk 4]
```diff
@@ -61,6 +61,7 @@ type BlackduckSpec struct {
 	ScanType              string                    `json:"scanType,omitempty"`
 	PersistentStorage     bool                      `json:"persistentStorage"`
 	PVC                   []PVC                     `json:"pvc,omitempty"`
+	NodeAffinities        []NodeAffinity            `json:"nodeAffinities,omitempty"`
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/api/blackduck/v1/types.go]: Good idea, done.

File: pkg/api/blackduck/v1/types.go [hunk 5]
```diff
@@ -92,6 +93,16 @@ type PVC struct {
 	StorageClass string `json:"storageClass,omitempty"`
 }
 
+// NodeAffinity will contain the specifications of a node affinity
+// TODO: currently, keeping it simple, but can be modified in the future to take in complex scenarios
+type NodeAffinity struct {
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/api/blackduck/v1/types.go]: Good idea, done.

File: pkg/api/blackduck/v1/types.go [hunk 6]
```diff
@@ -92,6 +93,16 @@ type PVC struct {
...
+	PodName      string `json:"podName"`
+	AffinityType string `json:"affinityType"`
+	Key          string `json:"key"`
+	Op           string `json:"op"`
+	Value        string `json:"value"`
```
[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/api/blackduck/v1/types.go]: Multiple definitions is fine for now.  I didn't want to make the parsing too complex for now.

[the remark this answers is not in the document]
Comment (rrati) [reply pkg/api/blackduck/v1/types.go]: I'm not sure I agree.  Changing this from a single string to an array later is a breaking change we should avoid.  Making it an array now shouldn't cause too much issue in the parsing should it?

[the remark this answers is not in the document]
Comment (yashbhutwala) [reply pkg/api/blackduck/v1/types.go]: Done.

**Comment by yashbhutwala** [comment]
All comments integrated, changed how crdupdater resolves the affinity change, and tested a few edge cases.  @rrati ready for another round of review, will squash all commits once you give it a green light.

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/413#pullrequestreview-247239645" title="LGTM">rrati</a>*, *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/413#" title="Author self-approved">yashbhutwala</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [rrati,yashbhutwala]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**